Hotness Score for Tags Should Use Article positive_reactions_count

mstruve opened the pull request:
## What type of PR is this? (check all applicable)
- [x] Bug Fix

## Description
Through some investigative work, I found that [we are not using reactions_count](https://github.com/thepracticaldev/dev.to/pull/6781#issuecomment-602835034) for articles or comments but instead we only [update `postive_reactions_count`](https://github.com/thepracticaldev/dev.to/blob/master/app/models/concerns/reactable.rb#L9) therefore that is the value we should be used when calculating the Tag hotness score. I included a data update job to fix all Tag hotness scores.

## Added tests?
- [x] yes

![alt_text](https://media3.giphy.com/media/1BdIPQfBbxb4gP7JFw/giphy.gif)

rhymes commented on spec/lib/data_update_scripts/update_tag_hotness_scores_spec.rb:
```diff
@@ -0,0 +1,8 @@
+require "rails_helper"
```
maybe we can skip this test as we're testing Rails in reality

mstruve replied:
I was using it to ensure we still had a Tag AR model in the event that we remove it, I know HIGHLY unlikely. I'm game for removing it to save time. I added it to try and be "responsible" and test my new script 😉

rhymes replied:
ahha our migrations and seed file would probably fail as well if that happens :D

mstruve commented:
This will definitely shift tags around in all of those places mentioned @rhymes but if I remember correctly, a little bit ago you mentioned that hotness scores for Tags seemed off when you were looking at some tags. I think this will put tags more inline with what people expect but it definitely will shift somethings around. I tagged @maestromac and @benhalpern in hopes of getting their thoughts on this "fix"

Another thought is we could minimize the effect and remove the multiplier to start. Not sure how much that would change things but probably a bit.

maestromac commented on app/models/tag.rb:
```diff
@@ -87,7 +87,7 @@ def calculate_hotness_score
       where("articles.featured_number > ?", 7.days.ago.to_i).
       map do |article|
-        (article.comments_count * 14) + (article.reactions_count * 4) + rand(6) + ((taggings_count + 1) / 2)
+        (article.comments_count * 14) + (article.positive_reactions_count * 4) + rand(6) + ((taggings_count + 1) / 2)
```
Just thinking out loud here. I think the multiplier could be reduced a bit as reaction count fluctuate more than comment count. Looks like we're calculating based on articles featured in the recent week so it's not going to be a gigantic change. Maybe we could simplify this by not giving any weights (ie `article.comments_count + article.positive_reactions_count`)? What's your thought @benhalpern ?

A bit off topic but I'm wondering how might these affect the formula too.
- What is the purpose of `rand(6)`? a dice roll?
- Why `taggings_count + 1` ?

benhalpern commented:
Yeah, this score calc is pretty arbitrary, so if we want to illiminate multiplication that's probably fine, and/or do +1 to avoid zeros.

FWIW, and we should document this, I think we should ideally now actually be using `score` in these contexts, which wasn't always a column, because if someone manipulates `positive_reactions_count`, a mod might come in and vomit it, which will offset that, but it still might have the bumped score.

benhalpern commented on lib/data_update_scripts/20200324133751_update_tag_hotness_scores.rb:
```diff
@@ -0,0 +1,8 @@
+  class UpdateTagHotnessScores
+    def run
+      # Saving a tag will trigger calculate_hotness_score
+      Tag.find_each(&:save)
```
IMO this proposed change is fairly mild in its consequences and we probably don't need to re-save as opposed to letting it run its course naturally.

That's just a thought, I don't know if we _always_ need these scripts on small scoring changes.

mstruve commented:
> IMO this proposed change is fairly mild in its consequences 

I think some tag hotness scores are going to change dramatically for those attached to articles with a lot of reactions which is why I added this. 

I removed the multiplication for now to downplay the impact of this change. I will leave a revamp/documentation of this scoring to someone else who really wants to get in the weeds and figure out the best approach. For now I just want to make sure tags linked to popular articles get the boost they deserve.

benhalpern commented:
Hmmmmm yeah @mstruve you're right. I was under the impression that `positive_reactions_count` and `reactions_count` would be almost always the same, or off by just a couple, but actually it looks like our `counter_culture` logic is flawed and we actually never even record `reactions_count` so they're all zero.

So yeah, script makes sense. But also I think `score` is probably the better number in places like this all in all.

mstruve commented:
So do we want to push this through as is or @benhalpern do you want to tweak that score method more?

benhalpern commented on app/models/tag.rb:
```diff
@@ -87,7 +87,7 @@ def calculate_hotness_score
       where("articles.featured_number > ?", 7.days.ago.to_i).
       map do |article|
-        (article.comments_count * 14) + (article.reactions_count * 4) + rand(6) + ((taggings_count + 1) / 2)
+        (article.comments_count * 14) + article.positive_reactions_count + rand(6) + ((taggings_count + 1) / 2)
```
```suggestion
        (article.comments_count * 14) + article.score + rand(6) + ((taggings_count + 1) / 2)
```

I'm good to go with as is, but this is all I was proposing. `score` is basically the sum of positive and negative reactions on  a post.
